File: src/main/java/com/soapdogg/lispInterpreter/parser/ExpressionNodeFinisher.java
package com.soapdogg.lispInterpreter.parser;

import com.soapdogg.lispInterpreter.asserter.TokenKindAsserter;
import com.soapdogg.lispInterpreter.datamodels.ParserResult;
import com.soapdogg.lispInterpreter.datamodels.Token;
import lombok.AllArgsConstructor;

import java.util.List;
import java.util.Optional;

@AllArgsConstructor(staticName = "newInstance")
public class ExpressionNodeFinisher {

    private final TokenKindAsserter tokenKindAsserter;
    private final ParserResultBuilder parserResultBuilder;

    public ParserResult finishParsingExpressionNode(
        final ParserResult result
    ) throws Exception {
        final List<Token> remainingTokens = result.getRemainingTokens();
        final Optional<Token> closeTokenOptional = Optional.ofNullable(remainingTokens.remove(0));
        final Token closeToken = tokenKindAsserter.assertTokenIsNotNull(closeTokenOptional);
        tokenKindAsserter.assertTokenIsClose(closeToken);
        return parserResultBuilder.buildParserResult(
            result.getResultingNode(),
            remainingTokens
        );
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/parser/ExpressionNodeFinisher.kt
package com.soapdogg.lispInterpreter.parser

import com.soapdogg.lispInterpreter.datamodels.ParserResult

class ExpressionNodeFinisher(
    private val parserResultBuilder: ParserResultBuilder
) {

    fun finishParsingExpressionNode(
        result: ParserResult
    ): ParserResult {
        return parserResultBuilder.buildParserResult(
            result.resultingNode,
            result.remainingTokens.subList(1, result.remainingTokens.size)
        )
    }
}